### ISSUE TITLE
Refactor Particle struct

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#2239]
Objective: reduce memory footprint of particle lists during `short_range_loop()` execution
- [ ] **re-shuffle Particle struct into dedicated structs or classes**
  + group members relevant to `short_range_loop()` in the same struct
  + refactor all functions processing Particle structs (member names should already be unique)
- [x] **write benchmarks for different projects**: magnetic fluids, ...
- [x] harmonize Python and C++ interface
  +  use same names for particle properties in C++ and Python?
  + be able to change the particle properties directly in Python, if possible
- [x] make it easy to add new particle properties in the relevant structs

Possible implementation:
```c++
struct Particle {
  ParticleProperties* p; /* data that has to be known for all ghosts, doesn't change */
  ParticleLocal* l; /* data needed only on the node the particle belongs to */
  ParticlePosition* r; /* positional data */
  ParticleMomentum* m;
  ParticleForce* f;
};
```
Rudolf's proposition to move members to dedicated structs:
- ParticleProperties:
  * Shan-Chen -> ParticleLocal?
  * affinity: probably part of object in fluid. Can it go to ParticleLocal?
  * membrane collision -> ParticleLocal?
  * rotation -> ParticleLocal?
  * electrohydrodynamic mu_e -> ParticleLocal
  * ext_force -> ParticleLocal
- ParticlePosition:
  * remove quatu (director)
  * remove dip
  * bond constraint: p_old -> ParticleLocal
  * Shan-Chen fluid composition -> ParticleLocal
- Particle:
  * can/should the bond and exclusion lists go to ParticleProperties?
  * swimming properties -> ParticleProperties or ParticleLocal

**Comment by hmenke** [comment]
I think that's a bad idea because older parts of Espresso might rely on the exact order of the data members (for strided `memcpy` an such).

**Comment by fweik** [comment]
@hmenke I don't think so. Do you have something specific in mind? Also such code should be fixed/removed anyhow...

**Comment by pdebuyl** [comment]
The only occurence of memcpy in the current python branch is to actually error out on the use of memcpy for the particles.

https://github.com/espressomd/espresso/blob/python/src/core/particle_data.hpp#L412

**Comment by hmenke** [comment]
@fweik That is the case for swimmers in LB GPU.  It's also the most efficient way to perform strided `memcpy`.

https://github.com/espressomd/espresso/blob/8ae11ece4bff74c4bbfac726148acdb2c521893d/src/core/cuda_interface.hpp#L46-L58

**Comment by fweik** [comment]
@hmenke I don't get your point. Can you make an example of something that is possible now and would not be after the change? Even now particles in Espresso are not trivially copyable.

**Comment by hmenke** [comment]
@fweik Apparently this is not an issue for `Particle` as Pierre confirmed, just something to keep in mind when refactoring memory layouts.

**Comment by RudolfWeeber** [comment]
Benchmarks:
Based on our discussion two weeks ago, I think, the following is reasonable
* Dilute and dense LJ liquid:
  * 1k and 10k particles per processor core
  * 1,2,8,16 cores
* Neutral liquid of charged lj particles:
  * Charges of +,-1, neutral system
  * P3M at 1E-4 accuracy
  * Same sizes and cores as for lj liquid

* Each individual simulation should run about 1-2 minutes to limit noise
* Tests should be run once with minimal (LJ+ELECTROSTATICS), once with 
  default and once with maintainer/configs/maxset.hpp as myconfig.hpp
* Equilibration integration needs to be run before measuring
* The skin needs to be tuned via cell_system.tune_skin() before measuring
* An exclusively allocated node on bee can be used as reference hardware for the time being


Goals of the particle structure optimizations:
* improved performance on the individual benchmarks
* marked reduction in the performance impact of adding more features to myconfig.hpp, particularly those that add data members to Particle and its substructs.

**Comment by jngrad** [comment]
Most of the tasks have been addressed between releases 4.0.0 and 4.1.4. The short range loop task is no longer up-to-date given how much has changed in the `Particle` struct. The core team is currently looking at other ways to improve performance, such as automatic SoA to AoS conversion or pulling the `ParticleForce` struct out of the `Particle` struct to facilitate reduction operations.
